include/ginkgo/core/base/types.hpp

```diff
@@ -193,6 +194,147 @@ using default_precision = double;
+{
+    return 0;
+}
+
+/**
+ * @copydoc shift(const std::array<char, num_groups>)
+ *
+ * @note this is the usual case of nested template
+ */
+template <int num_groups, int current_shift>
+constexpr std::enable_if_t<(num_groups > current_shift + 1), int> shift(
+    const std::array<char, num_groups> &bits)
+{
+    return bits[current_shift + 1] +
+           shift<num_groups, (current_shift + 1)>(bits);
+}
+
+
+}  // namespace detail
+
+
+// a shortcut type for ConfigSet
+using ConfigSetType = unsigned int;
+
+/**
+ * ConfigSet is a way to embed several information into one integer by given
+ * certain bits. The usage will be the following
+ * Set the method with bits Cfg = ConfigSet<nb_1, nb_2, ..., nb_k>
+ * Encode the given infomation encoded = Cfg::encode(i_1, i_2, ..., i_k)
+ * Decode the specific position information i_t = Cfg::decode<t>(encoded)
+ * The encoded result will use 32 bits to record
+ * rrrrr1..12....2...k..k, which 1/2/k means the bits store the information for
+ * 1/2/k position and r is for rest of unused bits.
+ *
+ * @tparam num_bits...  the number of bits for each position.
+ *
+ * @note the num_bit is required at least $log_2(maxval) + 1$
+ */
+template <int... num_bits>
+class ConfigSet {
+public:
+    static constexpr size_type num_groups = sizeof...(num_bits);
+    static constexpr std::array<char, num_groups> bits{num_bits...};
+
+    /**
+     * Decodes the `position` information from encoded
+     *
+     * @tparam position  the position of desired information
+     *
+     * @param encoded  the encoded integer
+     *
+     * @return the decoded information at position
+     */
+    template <int position>
+    static constexpr ConfigSetType decode(ConfigSetType encoded)
+    {
+        static_assert(position < num_groups,
+                      "This position is over the bounds.");
+        constexpr int shift = detail::shift<num_groups, position>(bits);
+        constexpr auto mask = detail::mask<bits[position]>();
+        return (encoded >> shift) & mask;
+    }
+
+    /**
+     * Encodes the information with given bit set to encoded integer.
+     *
+     * @note the last case of nested template.
+     */
+    template <size_type current_iter>
+    static constexpr std::enable_if_t<(current_iter == num_groups),
+                                      ConfigSetType>
+    encode()
+    {
+        return 0;
+    }
+
+    /**
+     * Encodes the information with given bit set to encoded integer.
+     *
+     * @tparam current_iter  the encoded place
+     * @tparam Rest...  the rest type
+     *
+     * @param first  the current encoded information
+     * @param rest...  the rest of others information waits for encoding
+     *
+     * @return the encoded integer
+     */
+    template <size_type current_iter = 0, typename... Rest>
+    static constexpr std::enable_if_t<(current_iter < num_groups),
+                                      ConfigSetType>
+    encode(ConfigSetType first, Rest &&... rest)
+    {
```

Comment:
we use this for runtime and compile time, so first is not guaranteed to be compile-time arguments.
`static_assert` requires all thing compile-time known.
`assert` limits the function into runtime only.
If we need the check, we need to distinguish runtime and compile time function.
compile time will use `ConfigSet::encode<...>`, which uses static_assert
runtime will use `ConfigSet::encode(...)`, which uses assert

Reply:
Maybe a hand-written constexpr assert function? It's fine if you think it's unnecessary. Maybe you could just create an issue to record it.

Reply:
what do you mean hand-written constexpr assert?
putting constexpr function into static_assert or write a constexpr check and throw?
the first one I tried but fail for the same reason.
the second one I need to check throw in constexpr is allowable or not.

Reply:
Yes, that does seem difficult. However, it looks like throwing from constexpr functions is mostly allowed? https://tech.jocodoma.com/2019/03/07/Exceptions-in-constexpr-Functions-in-CPP/

Reply:
it works.
throw will gives a compile time error from it is not constexpr function, which means the string can not give variable information.
throw in runtime is as usual.
some related code
https://godbolt.org/z/oa5d4arMs

Reply:
Excellent. We can use that.

Reply:
I would not rely on this. This seems like UB to me. See this other setting where you get exactly the opposite behavior:
https://stackoverflow.com/questions/34280729/throw-in-constexpr-function

Reply:
I saw that and my tentative conclusion was that it can be used, though I am no expert. I guess "ill-formed" != UB. In any case, if this works with all our CI toolchains, I vote to keep it because it offers some value.

Reply:
If the constexpr hit the condition, then the compiler will add the throw into constexpr.
However, it is impossible to put the throw in core constexpr, so the compiler error is the throw is not const statement or something.

I guess if the compiler does not ignore the condition in compile time, the compile error is always happened.

For non-constexpr, it just perform as usual.

Reply:
I try several gcc compilers on compiler explorer.
It is failed with gcc version <= 5.x no matter hit the condition or not.
I think 5.4 is a default compiler in Ubuntu 16.04 and the CUDA 9.2/10.x also allow this compiler version.